### PR TITLE
WL-2063: Prevent the URL to be shortened when it isn't necessary

### PR DESCRIPTION
**Pull request by ColinHebert**

### CODE REVIEW

File: kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java [hunk 1]
```diff
@@ -328,23 +328,11 @@ private String processUrl(ContentCopyContext context, String value,
 		try {
 			URI uri = new URI(value);
 			uri = uri.normalize();
-			if ("http".equals(uri.getScheme())
-					|| "https".equals(uri.getScheme())) {
-				if (uri.getHost() != null) {
-					if (servers.contains(uri.getHost())) {
-						// Drop the protocol and the host.
```
Comment (buckett) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java:335]: What's the reasoning behind keeping the protocol/hostname now?

Comment (ColinHebert) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java:335]: When an URL is shortened, a unique hash is generated based on the original URL, meaning that "site_id/relative/resource.pdf" and "http://weblearn.ox.ac.uk/site_id/relative/resource.pdf" are completely different and generate two different short URLs.

---

Say that `1234` is the hash for `http://weblearn.ox.ac.uk/site_id/relative/resource.pdf`, `abcd` is the hash for `second_site_id/relative/resource.pdf` and `9876` is the hash for `http://weblearn.ox.ac.uk/second_site_id/relative/resource.pdf`

Here is how the conversion of `<a href="http://weblearn.ox.ac.uk/x/1234">http://weblearn.ox.ac.uk/x/1234</a>` would work:

```
// 1/ replace on HTML attributes
     <a href="http://weblearn.ox.ac.uk/x/1234">http://weblearn.ox.ac.uk/x/1234</a>
// 1.1/ expand the URL if possible
     <a href="http://weblearn.ox.ac.uk/site_id/relative/resource.pdf">http://weblearn.ox.ac.uk/x/1234</a>
// 1.2/ replace the site_id with the new one (and with the current code, aslo reduce the URL)
     <a href="second_site_id/relative/resource.pdf">http://weblearn.ox.ac.uk/x/1234</a>
// 1.3/ re-shorten the URL
     <a href="http://weblearn.ox.ac.uk/x/abcd">http://weblearn.ox.ac.uk/x/1234</a>

// 2/ replace URLs with a site_id based on a UUID
// 2.1/ expand shortened URL
     <a href="http://weblearn.ox.ac.uk/x/abcd">http://weblearn.ox.ac.uk/site_id/relative/resource.pdf</a>
// 2.2/ replace the site_id with the new one
     <a href="http://weblearn.ox.ac.uk/x/abcd">http://weblearn.ox.ac.uk/second_site_id/relative/resource.pdf</a>
// 2.3/ re-shorten the URL
     <a href="http://weblearn.ox.ac.uk/x/abcd">http://weblearn.ox.ac.uk/x/9876</a>
```

There were two solutions, the first one (this one) is to limit the change to the URL to the site ID, the second solution would be trying to reduce the URL either before or after 2.2

---

To be honest, I'm not sure to understand what was the reasoning behind the removal of the protocol/hostname. I think it's doing a bit more than what is expected from the duplication of a site.

Comment (buckett) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java:335]: To me the fact that you can get a different hash for http://weblearn.ox.ac.uk/access/site/content.pdf and /access/site/content.pdf is a bug in the URL shortening service.

The protocol/hostname are superfluous and make it harder to test anything on a server other than the live server. If content has links such as <a href="/access/site/content.pdf">My Content</a> then they work on staging as well.

Comment (ColinHebert) [inline kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentCopyImpl.java:335]: I then would argue that the problem lies in the fact that people use absolute URL in their site.

This should be tackled by each tool accepting URLs, not by the URL shortener.

The fact is, a site you duplicate on live, then put on staging will have different behaviours depending on which site your using (the original one, or the copy).

**Pull request merged**